File: truevfs-kernel/truevfs-kernel-impl/src/main/scala/net/java/truevfs/kernel/impl/ArchiveControllerAdapter.scala
/*
 * Copyright (C) 2005-2015 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl

import java.{lang => jl, util => ju}

import net.java.truecommons.cio.Entry._
import net.java.truecommons.cio._
import net.java.truecommons.shed.BitField
import net.java.truevfs.kernel.spec._

private final class ArchiveControllerAdapter(override val getParent: FsController, controller: ArchiveController[_])
  extends FsAbstractController(controller.model) {

  override def node(options: AccessOptions, name: FsNodeName): FsNode = (controller node(options, name)).orNull

  override def checkAccess(options: AccessOptions, name: FsNodeName, types: BitField[Access]): Unit = {
    controller checkAccess(options, name, types)
  }

  override def setReadOnly(options: AccessOptions, name: FsNodeName): Unit = {
    controller setReadOnly(options, name)
  }

  override def setTime(options: AccessOptions, name: FsNodeName, times: ju.Map[Access, jl.Long]): Boolean = {
    controller setTime(options, name, times)
  }

  override def setTime(options: AccessOptions, name: FsNodeName, types: BitField[Access], value: Long): Boolean = {
    controller setTime(options, name, types, value)
  }

  def input(options: AccessOptions, name: FsNodeName): AnyInputSocket = controller input(options, name)

  def output(options: AccessOptions, name: FsNodeName, template: Entry): AnyOutputSocket = {
    controller output(options, name, Option(template))
  }

  override def make(options: AccessOptions, name: FsNodeName, tµpe: Type, template: Entry): Unit = {
    controller make(options, name, tµpe, Option(template))
  }

  override def unlink(options: AccessOptions, name: FsNodeName): Unit = {
    controller unlink(options, name)
  }

  override def sync(options: SyncOptions): Unit = controller sync options
}

File: truevfs-kernel/truevfs-kernel-impl/src/main/java/net/java/truevfs/kernel/impl/ArchiveControllerAdapter.java
/*
 * Copyright (C) 2005-2020 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl;

import net.java.truecommons.cio.Entry;
import net.java.truecommons.cio.InputSocket;
import net.java.truecommons.cio.OutputSocket;
import net.java.truecommons.shed.BitField;
import net.java.truevfs.kernel.spec.*;

import javax.annotation.CheckForNull;
import java.io.IOException;
import java.util.Map;
import java.util.Optional;

final class ArchiveControllerAdapter extends FsAbstractController {

    private final FsController parent;
    private final ArchiveController<?> controller;

    ArchiveControllerAdapter(FsController parent, ArchiveController<?> controller) {
        super(controller.getModel());
        this.parent = parent;
        this.controller = controller;
    }

    @Override
    public FsController getParent() {
        return parent;
    }

    @Override
    public FsNode node(BitField<FsAccessOption> options, FsNodeName name) throws IOException {
        return controller.node(options, name).orElse(null);
    }

    @Override
    public void checkAccess(BitField<FsAccessOption> options, FsNodeName name, BitField<Entry.Access> types) throws IOException {
        controller.checkAccess(options, name, types);
    }

    @Override
    public void setReadOnly(BitField<FsAccessOption> options, FsNodeName name) throws IOException {
        controller.setReadOnly(options, name);
    }

    @Override
    public boolean setTime(BitField<FsAccessOption> options, FsNodeName name, Map<Entry.Access, Long> times) throws IOException {
        return controller.setTime(options, name, times);
    }

    @Override
    public boolean setTime(BitField<FsAccessOption> options, FsNodeName name, BitField<Entry.Access> types, long value) throws IOException {
        return controller.setTime(options, name, types, value);
    }

    @Override
    public InputSocket<? extends Entry> input(BitField<FsAccessOption> options, FsNodeName name) {
        return controller.input(options, name);
    }

    @Override
    public OutputSocket<? extends Entry> output(BitField<FsAccessOption> options, FsNodeName name, @CheckForNull Entry template) {
        return controller.output(options, name, Optional.ofNullable(template));
    }

    @Override
    public void make(BitField<FsAccessOption> options, FsNodeName name, Entry.Type type, @CheckForNull Entry template) throws IOException {
        controller.make(options, name, type, Optional.ofNullable(template));
    }

    @Override
    public void unlink(BitField<FsAccessOption> options, FsNodeName name) throws IOException {
        controller.unlink(options, name);
    }

    @Override
    public void sync(BitField<FsSyncOption> options) throws FsSyncException {
        controller.sync(options);
    }
}
